trl/trainer/utils.py

```diff
@@ -173,6 +118,188 @@ def __call__(self, examples: list[dict[str, Any]]) -> dict[str, torch.Tensor]:
+            prompt_mask = prompt_mask[: len(input_ids)]
+
+        if input_ids[: len(prompt_ids)] != prompt_ids:
+            prompt_ids = input_ids[: len(prompt_ids)]
+            prompt_mask = [1] * len(prompt_ids)
+
+        labels = self._build_labels(input_ids, len(prompt_ids))
+
+        return {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask,
+            "labels": labels,
+            "prompt_input_ids": prompt_ids,
+            "prompt_attention_mask": prompt_mask,
+        }
+
+    def _build_labels(self, input_ids: list[int], prompt_length: int) -> list[int]:
+        labels = [self.ignore_index] * len(input_ids)
+        labels[prompt_length:] = input_ids[prompt_length:]
+        return labels
+
+    def _resolve_enable_thinking(
+        self, example: Mapping[str, Any], messages: Optional[Sequence[Mapping[str, Any]]], prompt: Optional[str]
+    ) -> Optional[bool]:
+        enable_thinking = self._extract_enable_thinking_metadata(example)
+        if enable_thinking is not None:
+            return enable_thinking
+
+        if prompt is not None:
+            inferred = self._infer_enable_thinking_from_prompt(prompt)
+            if inferred is not None:
+                return inferred
+
+        if messages:
+            first_message = messages[0] if messages and isinstance(messages[0], Mapping) else None
+            if first_message:
+                content = first_message.get("content")
+                if isinstance(content, str):
+                    if "/no_think" in content:
+                        return False
+                    if "/think" in content:
+                        return True
+        return None
+
+    def _apply_chat_template(
+        self, messages: Sequence[Mapping[str, Any]], *, add_generation_prompt: bool, enable_thinking: Optional[bool]
+    ) -> str:
+        kwargs: dict[str, Any] = {"tokenize": False, "add_generation_prompt": add_generation_prompt}
+        if enable_thinking is not None:
+            kwargs["enable_thinking"] = enable_thinking
+        return self.tokenizer.apply_chat_template(messages, **kwargs)
+
+    def _tokenize_text(self, text: str, max_length: Optional[int]) -> dict[str, list[int]]:
+        tokenized = self.tokenizer(
+            text,
+            truncation=True,
+            max_length=max_length if max_length is not None else self.max_length,
+            padding=False,
+            return_tensors=None,
+            add_special_tokens=False,
+        )
+        ids = tokenized["input_ids"]
+        mask = tokenized.get("attention_mask")
+        if mask is None:
+            mask = [1] * len(ids)
+        if max_length is not None and len(ids) > max_length:
+            ids = ids[:max_length]
+            mask = mask[:max_length]
+        return {"input_ids": ids, "attention_mask": mask}
+
+    def _extract_enable_thinking_metadata(self, example: Mapping[str, Any]) -> Optional[bool]:
```

Comment:
I think this method could also be made static:
```suggestion
    @staticmethod
    def _extract_enable_thinking_metadata(self, example: Mapping[str, Any]) -> Optional[bool]:
```